File: backend/src/dailyUpdate.js
/** The daily job: credit due monthly allowances, snapshot prices for every
 * held ticker, and record each wallet's total value for the day. Runs from
 * the cron trigger and the manual /admin/run-daily-update endpoint. */

import { getPriceWithFallback, membershipHoldings, valueHoldings } from "./valuation.js";
import { addMonths, r2, todayISO } from "./util.js";

export async function creditDueAllowances(env) {
  const { results: memberships } = await env.DB.prepare(
    `SELECT m.*, g.monthly_allowance FROM memberships m
     JOIN groups g ON g.id = m.group_id WHERE g.monthly_allowance > 0`
  ).all();

  const now = new Date();
  let credits = 0;
  for (const membership of memberships) {
    let last = new Date(membership.last_allowance_at);
    let cash = membership.cash_balance;
    const statements = [];
    while (addMonths(last, 1) <= now) {
      last = addMonths(last, 1);
      cash = r2(cash + membership.monthly_allowance);
      statements.push(
        env.DB.prepare(
          "INSERT INTO transactions (membership_id, type, amount, created_at) VALUES (?, 'ALLOWANCE', ?, ?)"
        ).bind(membership.id, membership.monthly_allowance, last.getTime())
      );
      credits++;
    }
    if (statements.length > 0) {
      statements.push(
        env.DB.prepare(
          "UPDATE memberships SET cash_balance = ?, last_allowance_at = ? WHERE id = ?"
        ).bind(cash, last.getTime(), membership.id)
      );
      await env.DB.batch(statements);
    }
  }
  return credits;
}

export async function runDailyUpdate(env) {
  const today = todayISO();
  const allowances = await creditDueAllowances(env);

  const { results: tickers } = await env.DB.prepare(
    "SELECT DISTINCT ticker FROM holdings"
  ).all();
  let pricesUpdated = 0;
  for (const { ticker } of tickers) {
    const price = await getPriceWithFallback(env, ticker);
    if (price === null) continue;
    await env.DB.prepare(
      `INSERT INTO price_snapshots (ticker, price, snapshot_date) VALUES (?, ?, ?)
       ON CONFLICT (ticker, snapshot_date) DO UPDATE SET price = excluded.price`
    ).bind(ticker, price, today).run();
    pricesUpdated++;
  }

  const { results: memberships } = await env.DB.prepare("SELECT * FROM memberships").all();
  for (const membership of memberships) {
    const holdings = await membershipHoldings(env, membership.id);
    const { total } = await valueHoldings(env, holdings);
    const totalValue = r2(membership.cash_balance + total);
    await env.DB.prepare(
      `INSERT INTO portfolio_snapshots (membership_id, total_value, cash_balance, snapshot_date)
       VALUES (?, ?, ?, ?)
       ON CONFLICT (membership_id, snapshot_date) DO UPDATE
         SET total_value = excluded.total_value, cash_balance = excluded.cash_balance`
    ).bind(membership.id, totalValue, membership.cash_balance, today).run();
  }

  return {
    date: today,
    allowances_credited: allowances,
    tickers_updated: pricesUpdated,
    portfolios_updated: memberships.length,
  };
}

File: backend/app/services/daily_update.py
"""The daily update job: credit allowances, snapshot prices for every held
ticker, and record each wallet's total value for the day."""

import logging
from datetime import date

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import Holding, Membership, PortfolioSnapshot, PriceSnapshot
from app.services import valuation
from app.services.allowance import credit_due_allowances

logger = logging.getLogger(__name__)


def run_daily_update(db: Session) -> dict:
    today = date.today()
    allowances = credit_due_allowances(db)
    prices = _snapshot_prices(db, today)
    portfolios = _snapshot_portfolios(db, today)
    db.commit()

    summary = {
        "date": today.isoformat(),
        "allowances_credited": allowances,
        "tickers_updated": prices,
        "portfolios_updated": portfolios,
    }
    logger.info("Daily update complete: %s", summary)
    return summary


def _snapshot_prices(db: Session, today: date) -> int:
    tickers = db.scalars(select(Holding.ticker).distinct()).all()
    updated = 0
    for ticker in tickers:
        price = valuation.get_price_with_fallback(db, ticker)
        if price is None:
            logger.warning("No price available for %s, skipping snapshot", ticker)
            continue
        existing = db.scalars(
            select(PriceSnapshot).where(
                PriceSnapshot.ticker == ticker, PriceSnapshot.snapshot_date == today
            )
        ).first()
        if existing:
            existing.price = price
        else:
            db.add(PriceSnapshot(ticker=ticker, price=price, snapshot_date=today))
        updated += 1
    return updated


def _snapshot_portfolios(db: Session, today: date) -> int:
    memberships = db.scalars(select(Membership)).all()
    for membership in memberships:
        total = valuation.get_total_value(db, membership)
        existing = db.scalars(
            select(PortfolioSnapshot).where(
                PortfolioSnapshot.membership_id == membership.id,
                PortfolioSnapshot.snapshot_date == today,
            )
        ).first()
        if existing:
            existing.total_value = total
            existing.cash_balance = membership.cash_balance
        else:
            db.add(
                PortfolioSnapshot(
                    membership_id=membership.id,
                    total_value=total,
                    cash_balance=membership.cash_balance,
                    snapshot_date=today,
                )
            )
    return len(memberships)
